Deprecation warnings on g++ 7.1.0 at MacOS 10.12.4

Not a big deal, everything working as intended, but you may want to look at these warnings when building code with Rcpp using g++ 7.1.0

```
In file included from /Library/Frameworks/R.framework/Versions/3.3/Resources/library/Rcpp/include/Rcpp/Module.h:338:0,
                 from /Library/Frameworks/R.framework/Versions/3.3/Resources/library/Rcpp/include/Rcpp.h:64,
                 from osutil.cpp:1:
/Library/Frameworks/R.framework/Versions/3.3/Resources/library/Rcpp/include/Rcpp/module/Module_Property.h:103:41: warning: dynamic exception specifications are deprecated in C++11 [-Wdeprecated]
     void set(Class* object, SEXP value) throw(std::range_error,Rcpp::not_compatible){
                                         ^~~~~
/Library/Frameworks/R.framework/Versions/3.3/Resources/library/Rcpp/include/Rcpp/module/Module_Property.h:161:41: warning: dynamic exception specifications are deprecated in C++11 [-Wdeprecated]
     void set(Class* object, SEXP value) throw(std::range_error,Rcpp::not_compatible){
                                         ^~~~~
```

I've just come across these warnings as well. However, while in C++11 and C++14 mode they're indeed **warnings**, in C++17 mode they're **errors** for me, making this quite a bit more pressing (since all you need to do to get these errors is to `#include <Rcpp.h>`, this renders Rcpp unusable in g++ in C++17 mode for now).
```
In file included from /usr/local/lib/R/3.4/site-library/Rcpp/include/Rcpp/Module.h:338:0,
                 from /usr/local/lib/R/3.4/site-library/Rcpp/include/Rcpp.h:64,
                 from foo.cc:1:
/usr/local/lib/R/3.4/site-library/Rcpp/include/Rcpp/module/Module_Property.h:103:41: error: ISO C++1z does not allow dynamic exception specifications
     void set(Class* object, SEXP value) throw(std::range_error,Rcpp::not_compatible){
                                         ^~~~~
/usr/local/lib/R/3.4/site-library/Rcpp/include/Rcpp/module/Module_Property.h:161:41: error: ISO C++1z does not allow dynamic exception specifications
     void set(Class* object, SEXP value) throw(std::range_error,Rcpp::not_compatible){
                                         ^~~~~
make: *** [foo.o] Error 1
g++-7 -I/usr/local/Cellar/r/3.4.0_1/R.framework/Resources/include -DNDEBUG   -I"/usr/local/lib/R/3.4/site-library/Rcpp/include" -I"/Users/pipping" -I/usr/local/opt/gettext/include -I/usr/local/opt/readline/include -I/usr/local/include  -std=c++17 -fPIC  -g -O2  -c foo.cc -o foo.o
```

According to [this thread on gcc-dev](http://gcc.1065356.n8.nabble.com/Disable-std-c-17-quot-ISO-C-1z-does-not-allow-dynamic-exception-specifications-quot-td1352263.html) there is no way to demote this error to a warning.

--- a/ChangeLog
+++ b/ChangeLog
@@ -2,6 +2,9 @@
 
         * DESCRIPTION (Version, Date): Roll minor version again
 
+        * inst/include/Rcpp/module/Module_Property.h: Remove two typed exception
+        specifications which upset g++ 7.1 or higher
+
 2017-05-05  Kirill Müller <krlmlr@mailbox.org>
 
         * inst/include/Rcpp/date_datetime/Date.h: Suppress -Wconversion warnings

--- a/inst/NEWS.Rd
+++ b/inst/NEWS.Rd
@@ -23,6 +23,9 @@
       collection (Dirk in \ghpr{684} addressing \ghit{683}).
       \item Several compiler warnings for \code{-Wconversions} are no longer
       generated (Kirill Mueller in \ghpr{687} and \ghpr{688})
+      \item Two exception specification that are no longer tolerated by
+      \code{g++-7.1} or later were removed (Dirk in \ghpr{690} addressing
+      \ghit{689})
     }
     \item Changes in Rcpp Documentation:
     \itemize{

--- a/inst/include/Rcpp/module/Module_Property.h
+++ b/inst/include/Rcpp/module/Module_Property.h
@@ -2,7 +2,7 @@
 //
 // Module_Property.h: Rcpp R/C++ interface class library -- Rcpp modules
 //
-// Copyright (C) 2010 - 2011 Dirk Eddelbuettel and Romain Francois
+// Copyright (C) 2010 - 2017  Dirk Eddelbuettel and Romain Francois
 //
 // This file is part of Rcpp.
 //
@@ -100,7 +100,7 @@ public:
     SEXP get(Class* object) {
         return Rcpp::wrap( (object->*getter)() ) ;
     }
-    void set(Class* object, SEXP value) throw(std::range_error,Rcpp::not_compatible){
+    void set(Class* object, SEXP value) {
         (object->*setter)(
                           Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
                            ) ;
@@ -158,7 +158,7 @@ public:
     SEXP get(Class* object) {
         return Rcpp::wrap( (object->*getter)() ) ;
     }
-    void set(Class* object, SEXP value) throw(std::range_error,Rcpp::not_compatible){
+    void set(Class* object, SEXP value) {
         setter( object,
                 Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
                 ) ;

--- a/inst/include/Rcpp/module/Module_Property.h
+++ b/inst/include/Rcpp/module/Module_Property.h
@@ -26,120 +26,116 @@
 template <typename Class, typename PROP>
 class CppProperty_GetMethod : public CppProperty<Class> {
 public:
-    typedef PROP (Class::*GetMethod)(void) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (Class::*GetMethod)(void);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetMethod( GetMethod getter_, const char* doc = 0 ) :
+    CppProperty_GetMethod(GetMethod getter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), class_name(DEMANGLE(PROP)){}
 
-    SEXP get(Class* object) { return Rcpp::wrap( (object->*getter)() ) ; }
-    void set(Class*, SEXP) { throw std::range_error("property is read only") ; }
-    bool is_readonly(){ return true ; }
+    SEXP get(Class* object) { return Rcpp::wrap((object->*getter)()); }
+    void set(Class*, SEXP) { throw std::range_error("property is read only"); }
+    bool is_readonly(){ return true; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    std::string class_name ;
+    GetMethod getter;
+    std::string class_name;
 
-} ;
+};
 
 // getter through a const member function
 template <typename Class, typename PROP>
 class CppProperty_GetConstMethod : public CppProperty<Class> {
 public:
-    typedef PROP (Class::*GetMethod)(void) const ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (Class::*GetMethod)(void) const;
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetConstMethod( GetMethod getter_ , const char* doc = 0) :
+    CppProperty_GetConstMethod(GetMethod getter_ , const char* doc = 0) :
         prop_class(doc), getter(getter_), class_name(DEMANGLE(PROP)){}
 
-    SEXP get(Class* object) { return Rcpp::wrap( (object->*getter)() ) ; }
-    void set(Class*, SEXP) { throw std::range_error("property is read only") ; }
-    bool is_readonly(){ return true ; }
+    SEXP get(Class* object) { return Rcpp::wrap((object->*getter)()); }
+    void set(Class*, SEXP) { throw std::range_error("property is read only"); }
+    bool is_readonly(){ return true; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    std::string class_name ;
+    GetMethod getter;
+    std::string class_name;
 
-} ;
+};
 
 
 // getter through a free function taking a pointer to Class
 template <typename Class, typename PROP>
 class CppProperty_GetPointerMethod : public CppProperty<Class> {
 public:
-    typedef PROP (*GetMethod)(Class*) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (*GetMethod)(Class*);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetPointerMethod( GetMethod getter_ , const char* doc = 0) :
+    CppProperty_GetPointerMethod(GetMethod getter_ , const char* doc = 0) :
         prop_class(doc), getter(getter_), class_name(DEMANGLE(PROP)){}
 
-    SEXP get(Class* object) { return Rcpp::wrap( getter(object) ) ; }
-    void set(Class*, SEXP) { throw std::range_error("property is read only") ; }
-    bool is_readonly(){ return true ; }
+    SEXP get(Class* object) { return Rcpp::wrap(getter(object)); }
+    void set(Class*, SEXP) { throw std::range_error("property is read only"); }
+    bool is_readonly(){ return true; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    std::string class_name ;
-} ;
+    GetMethod getter;
+    std::string class_name;
+};
 
 
 // getter and setter through member functions
 template <typename Class, typename PROP>
 class CppProperty_GetMethod_SetMethod : public CppProperty<Class> {
 public:
-    typedef PROP (Class::*GetMethod)(void) ;
-    typedef void (Class::*SetMethod)(PROP) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (Class::*GetMethod)(void);
+    typedef void (Class::*SetMethod)(PROP);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetMethod_SetMethod( GetMethod getter_, SetMethod setter_, const char* doc = 0) :
+    CppProperty_GetMethod_SetMethod(GetMethod getter_, SetMethod setter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), setter(setter_), class_name(DEMANGLE(PROP)){}
 
     SEXP get(Class* object) {
-        return Rcpp::wrap( (object->*getter)() ) ;
+        return Rcpp::wrap((object->*getter)());
     }
     void set(Class* object, SEXP value) {
-        (object->*setter)(
-                          Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
-                           ) ;
+        (object->*setter)(Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >(value));
     }
-    bool is_readonly(){ return false ; }
+    bool is_readonly(){ return false; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    SetMethod setter ;
-    std::string class_name ;
-} ;
+    GetMethod getter;
+    SetMethod setter;
+    std::string class_name;
+};
 template <typename Class, typename PROP>
 class CppProperty_GetConstMethod_SetMethod : public CppProperty<Class> {
 public:
-    typedef PROP (Class::*GetMethod)(void) const ;
-    typedef void (Class::*SetMethod)(PROP) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (Class::*GetMethod)(void) const;
+    typedef void (Class::*SetMethod)(PROP);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetConstMethod_SetMethod( GetMethod getter_, SetMethod setter_, const char* doc = 0) :
+    CppProperty_GetConstMethod_SetMethod(GetMethod getter_, SetMethod setter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), setter(setter_), class_name(DEMANGLE(PROP)){}
 
     SEXP get(Class* object) {
-        return Rcpp::wrap( (object->*getter)() ) ;
+        return Rcpp::wrap((object->*getter)());
     }
     void set(Class* object, SEXP value) {
-        (object->*setter)(
-                          Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
-                           ) ;
+        (object->*setter)(Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >(value));
     }
-    bool is_readonly(){ return false ; }
+    bool is_readonly(){ return false; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    SetMethod setter ;
-    std::string class_name ;
+    GetMethod getter;
+    SetMethod setter;
+    std::string class_name;
 
-} ;
+};
 
 
 
@@ -148,116 +144,108 @@ private:
 template <typename Class, typename PROP>
 class CppProperty_GetMethod_SetPointer : public CppProperty<Class> {
 public:
-    typedef PROP (Class::*GetMethod)(void) ;
-    typedef void (*SetMethod)(Class*,PROP) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (Class::*GetMethod)(void);
+    typedef void (*SetMethod)(Class*,PROP);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetMethod_SetPointer( GetMethod getter_, SetMethod setter_, const char* doc = 0) :
+    CppProperty_GetMethod_SetPointer(GetMethod getter_, SetMethod setter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), setter(setter_), class_name(DEMANGLE(PROP)){}
 
     SEXP get(Class* object) {
-        return Rcpp::wrap( (object->*getter)() ) ;
+        return Rcpp::wrap((object->*getter)());
     }
     void set(Class* object, SEXP value) {
-        setter( object,
-                Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
-                ) ;
+        setter(object, Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >(value));
     }
-    bool is_readonly(){ return false ; }
+    bool is_readonly(){ return false; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    SetMethod setter ;
-    std::string class_name ;
+    GetMethod getter;
+    SetMethod setter;
+    std::string class_name;
 
-} ;
+};
 template <typename Class, typename PROP>
 class CppProperty_GetConstMethod_SetPointer : public CppProperty<Class> {
 public:
-    typedef PROP (Class::*GetMethod)(void) const ;
-    typedef void (*SetMethod)(Class*,PROP) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (Class::*GetMethod)(void) const;
+    typedef void (*SetMethod)(Class*,PROP);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetConstMethod_SetPointer( GetMethod getter_, SetMethod setter_, const char* doc = 0) :
+    CppProperty_GetConstMethod_SetPointer(GetMethod getter_, SetMethod setter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), setter(setter_), class_name(DEMANGLE(PROP)){}
 
     SEXP get(Class* object) {
-        return Rcpp::wrap( (object->*getter)() ) ;
+        return Rcpp::wrap((object->*getter)());
     }
     void set(Class* object, SEXP value) {
-        setter( object,
-                Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
-                ) ;
+        setter(object, Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >(value));
     }
-    bool is_readonly(){ return false ; }
+    bool is_readonly(){ return false; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    SetMethod setter ;
-    std::string class_name ;
+    GetMethod getter;
+    SetMethod setter;
+    std::string class_name;
 
-} ;
+};
 
 // getter through pointer function, setter through member function
 template <typename Class, typename PROP>
 class CppProperty_GetPointer_SetMethod : public CppProperty<Class> {
 public:
-    typedef PROP (*GetMethod)(Class*) ;
-    typedef void (Class::*SetMethod)(PROP) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (*GetMethod)(Class*);
+    typedef void (Class::*SetMethod)(PROP);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetPointer_SetMethod( GetMethod getter_, SetMethod setter_, const char* doc = 0) :
+    CppProperty_GetPointer_SetMethod(GetMethod getter_, SetMethod setter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), setter(setter_), class_name(DEMANGLE(PROP)){}
 
     SEXP get(Class* object) {
-        return Rcpp::wrap( getter(object) ) ;
+        return Rcpp::wrap(getter(object));
     }
     void set(Class* object, SEXP value) {
-        (object->*setter)(
-                          Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
-                           ) ;
+        (object->*setter)(Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >(value));
     }
-    bool is_readonly(){ return false ; }
+    bool is_readonly(){ return false; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    SetMethod setter ;
-    std::string class_name ;
+    GetMethod getter;
+    SetMethod setter;
+    std::string class_name;
 
-} ;
+};
 
 // getter and setter through pointer functions
 // getter through pointer function, setter through member function
 template <typename Class, typename PROP>
 class CppProperty_GetPointer_SetPointer : public CppProperty<Class> {
 public:
-    typedef PROP (*GetMethod)(Class*) ;
-    typedef void (*SetMethod)(Class*,PROP) ;
-    typedef CppProperty<Class> prop_class ;
+    typedef PROP (*GetMethod)(Class*);
+    typedef void (*SetMethod)(Class*,PROP);
+    typedef CppProperty<Class> prop_class;
 
-    CppProperty_GetPointer_SetPointer( GetMethod getter_, SetMethod setter_, const char* doc = 0) :
+    CppProperty_GetPointer_SetPointer(GetMethod getter_, SetMethod setter_, const char* doc = 0) :
         prop_class(doc), getter(getter_), setter(setter_), class_name(DEMANGLE(PROP)){}
 
     SEXP get(Class* object) {
-        return Rcpp::wrap( getter(object) ) ;
+        return Rcpp::wrap(getter(object));
     }
     void set(Class* object, SEXP value) {
-        setter( object,
-                Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >( value )
-                ) ;
+        setter(object, Rcpp::as< typename Rcpp::traits::remove_const_and_reference< PROP >::type >(value));
     }
-    bool is_readonly(){ return false ; }
+    bool is_readonly(){ return false; }
     std::string get_class(){ return class_name; }
 
 private:
-    GetMethod getter ;
-    SetMethod setter ;
-    std::string class_name ;
+    GetMethod getter;
+    SetMethod setter;
+    std::string class_name;
 
-} ;
+};
 
 
 #endif

